staging/src/k8s.io/client-go/rest/with_retry.go

```diff
@@ -147,34 +150,88 @@ func (r *withRetry) NextRetry(req *http.Request, resp *http.Response, err error,
 	//     need to check if it is retryable
 	seconds, wait := checkWait(resp)
 	if !wait {
-		return retryAfter, false
+		return false
 	}
 
-	retryAfter.Wait = time.Duration(seconds) * time.Second
-	retryAfter.Reason = getRetryReason(r.attempts, seconds, resp, err)
-	return retryAfter, true
+	r.retryAfter.Wait = time.Duration(seconds) * time.Second
+	r.retryAfter.Reason = getRetryReason(r.attempts, seconds, resp, err)
+
+	if err := r.prepareForNextRetry(ctx, restReq); err != nil {
+		klog.V(4).Infof("Could not retry request - %v", err)
+		return false
+	}
+
+	return true
 }
 
-func (r *withRetry) BeforeNextRetry(ctx context.Context, backoff BackoffManager, retryAfter *RetryAfter, url string, body io.Reader) error {
+// prepareForNextRetry is responsible for carrying out operations that need
+// to be completed before the next retry is initiated:
+// - if the request context is already canceled there is no need to
+//   retry, the function will return ctx.Err().
+// - we need to seek to the beginning of the request body before we
+//   initiate the next retry, the function should return an error if
+//   it fails to do so.
+func (r *withRetry) prepareForNextRetry(ctx context.Context, request *Request) error {
 	// Ensure the response body is fully read and closed before
 	// we reconnect, so that we reuse the same TCP connection.
 	if ctx.Err() != nil {
 		return ctx.Err()
 	}
 
-	if seeker, ok := body.(io.Seeker); ok && body != nil {
+	if seeker, ok := request.body.(io.Seeker); ok && request.body != nil {
 		if _, err := seeker.Seek(0, 0); err != nil {
-			return fmt.Errorf("can't Seek() back to beginning of body for %T", r)
+			return fmt.Errorf("can't Seek() back to beginning of body for %T", request)
 		}
 	}
 
-	klog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", retryAfter.Wait, retryAfter.Attempt, url)
-	if backoff != nil {
-		backoff.Sleep(retryAfter.Wait)
+	klog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", r.retryAfter.Wait, r.retryAfter.Attempt, request.URL().String())
+	return nil
+}
+
+func (r *withRetry) Before(ctx context.Context, request *Request) error {
+	// r.retryAfter represents the retry after parameters for the
+	// next attempt to be made immediately after this function returns.
+	url := request.URL()
```

Comment:
it would be cleaner to follow if we would change to:
```
if r.retryAfter == nil {
  return nil
}

if request.backoff != nil {
   ...
}
...
```

Reply:
done